File: benches/lang-comparison/go/linked_list.go
// Singly linked list with basic operations
package main

import "fmt"

type Node struct {
	Value int64
	Next  *Node
}

func newNode(val int64) *Node {
	return &Node{Value: val, Next: nil}
}

func listPush(head *Node, val int64) {
	n := newNode(val)
	n.Next = head.Next
	head.Next = n
}

func listLen(head *Node) int64 {
	var count int64
	cur := head.Next
	for cur != nil {
		count++
		cur = cur.Next
	}
	return count
}

func listSum(head *Node) int64 {
	var total int64
	cur := head.Next
	for cur != nil {
		total += cur.Value
		cur = cur.Next
	}
	return total
}

func main() {
	head := &Node{Value: 0, Next: nil}
	for i := int64(1); i <= 10; i++ {
		listPush(head, i)
	}
	fmt.Println(listLen(head))
	fmt.Println(listSum(head))
}

File: benches/lang-comparison/python/linked_list.py
# Singly linked list with basic operations
from dataclasses import dataclass
from typing import Optional

@dataclass
class Node:
    value: int
    next: Optional["Node"] = None

def list_push(head: Node, val: int) -> None:
    new_node = Node(value=val, next=head.next)
    head.next = new_node

def list_len(head: Node) -> int:
    count = 0
    cur = head.next
    while cur is not None:
        count += 1
        cur = cur.next
    return count

def list_sum(head: Node) -> int:
    total = 0
    cur = head.next
    while cur is not None:
        total += cur.value
        cur = cur.next
    return total

def main():
    head = Node(value=0)
    for i in range(1, 11):
        list_push(head, i)
    print(list_len(head))
    print(list_sum(head))

if __name__ == "__main__":
    main()
